lib/policies/timestamp-generation.js

```diff
@@ -0,0 +1,152 @@
+ * @abstract
+ */
+TimestampGenerator.prototype.next = function (client) {
+  throw new Error('next() must be implemented');
+};
+
+/**
+ * A timestamp generator that guarantees monotonically increasing timestamps and logs warnings when timestamps
+ * drift in the future.
+ * <p>
+ *   {@link Date} has millisecond precision and client timestamps require microsecond precision. This generator
+ *   keeps track of the last generated timestamp, and if the current time is within the same millisecond as the last,
+ *   it fills the microsecond portion of the new timestamp with the value of an incrementing counter.
+ * </p>
+ * @param {Number} [warningThreshold] Determines how far in the future timestamps are allowed to drift before a
+ * warning is logged, expressed in milliseconds. Default: <code>1000</code>.
+ * @param {Number} [minLogInterval] In case of multiple log events, it determines the time separation between log
+ * events, expressed in milliseconds. Use 0 to disable. Default: <code>1000</code>.
+ * @extends {TimestampGenerator}
+ * @constructor
+ */
+function MonotonicTimestampGenerator(warningThreshold, minLogInterval) {
+  if (warningThreshold < 0) {
+    throw new errors.ArgumentError('warningThreshold can not be lower than 0');
+  }
+  this._warningThreshold = warningThreshold || 1000;
+  this._minLogInterval = 1000;
+  if (typeof minLogInterval === 'number') {
+    // A value under 1 will disable logging
+    this._minLogInterval = minLogInterval;
+  }
+  this._micros = -1;
+  this._lastDate = 0;
+  this._lastLogDate = 0;
+}
+
+util.inherits(MonotonicTimestampGenerator, TimestampGenerator);
+
+/**
+ * Returns the current time in milliseconds since UNIX epoch
+ * @returns {Number}
+ */
+MonotonicTimestampGenerator.prototype.getDate = function () {
+  return Date.now();
+};
+
+MonotonicTimestampGenerator.prototype.next = function (client) {
+  var date = this.getDate();
+  var drifted = 0;
+  if (date > this._lastDate) {
+    this._micros = 0;
+    this._lastDate = date;
+    return this._generateMicroseconds();
+  }
+
+  if (date < this._lastDate) {
+    drifted = this._lastDate - date;
+    date = this._lastDate;
+  }
+  if (++this._micros === 1000) {
+    this._micros = 0;
+    if (date === this._lastDate) {
+      // Move date 1 millisecond into the future
+      date++;
+      drifted++;
+    }
+  }
+  if (drifted >= this._warningThreshold) {
+    // Avoid logging an unbounded amount of times within a clock-skew event or during an interval when more than 1
+    // query is being issued by microsecond
+    var currentLogDate = Date.now();
+    if (this._minLogInterval > 0 && this._lastLogDate + this._minLogInterval <= currentLogDate){
+      var message = util.format(
+        'Timestamp generated using current date was %d milliseconds behind the last generated timestamp (%d),' +
+        'returned value (%d) is being artificially incremented to guarantee monotonicity.',
+        drifted, this._lastDate, date);
```

Comment:
Another thing I realized is that during a leap second (a normal case where this would happen) is that the interval between generating timestamps is probably > 0ms, so a warning would not be emitted during a leap second.

However, I think setting the interval lower than 1 second in practice is probably not a good idea, especially since ntp might adjust your clock pretty frequently in virtualized environments, so maybe we just leave it as it.